### PR TITLE
drivers: sensor: Added driver for the Würth Elektronik WSEN-TIDS sensor

### PR DESCRIPTION
**Pull request by mah-eiSmart**
This pull request contains the driver and a sample for the Würth Elektronik WSEN-TIDS sensor

### CODE REVIEW

**Comment by mah-eiSmart** [comment]
Hello @MaureenHelm ,
may I ask for a review?
The content of the code is similar to the previous contributed HIDS sensor driver.
Thx,
Matthias

File: drivers/sensor/wsen_tids/wsen_tids.c [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+	100,
+	200,
+};
+
+static int tids_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct tids_data *data = dev->data;
+	int16_t raw_temperature;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids.c:35]: Return an errno

File: drivers/sensor/wsen_tids/wsen_tids.c [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	const struct tids_config *const config = dev->config;
+	struct tids_data *data = dev->data;
+	int status;
+	uint8_t device_id;
+	struct sensor_value odr;
+
+	/* Initialize WE sensor interface */
+	TIDS_getDefaultInterface(&data->sensor_interface);
+	data->sensor_interface.interfaceType = WE_i2c;
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids.c:136]: Is this sensor capable of supporting anything other than I2C? If not, then I'd remove the conditionals

File: drivers/sensor/wsen_tids/wsen_tids.c [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+	if (device_id != TIDS_DEVICE_ID_VALUE) {
+		LOG_ERR("Invalid device ID 0x%x.", device_id);
+		return -EIO;
+	}
+
+	/* Reset sensor */
+	TIDS_softReset(&data->sensor_interface, TIDS_enable);
+	k_sleep(K_USEC(5));
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids.c:153]: Where does this value come from? Is it in the datasheet?

File: drivers/sensor/wsen_tids/wsen_tids_trigger.c [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+	tids_process_threshold_interrupt(tids->dev);
+}
+#endif /* CONFIG_WSEN_TIDS_TRIGGER_GLOBAL_THREAD */
+
+int tids_threshold_set(const struct device *dev, const struct sensor_value *thresh_value,
+		       bool upper)
+{
+	struct tids_data *data = dev->data;
+	double thresh = (sensor_value_to_double(thresh_value) / 0.64) + 63;
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids_trigger.c:140]: Please document these magic numbers with comments or macros

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
I'd like to stop adding driver-specific sample applications. Can you look at adding a generic temperature sensor application and enabling it for this sensor, similar to #47005 and #47944?

**Comment by mah-eiSmart** [comment]
Hello @MaureenHelm ,
thanks for the review. I handled the commented points.
Concerning an generic example, I talked with my manager. He won't give me the time NOW, to generate a generic temperature sensor example. In 2023 maybe.
Thus, in case you still not accept the example, I just push the driver and will generate an example later in a separate pull request.
What do you think?
Thank you, and best regards
Matthias

**Comment by MaureenHelm** [comment]
> Thus, in case you still not accept the example, I just push the driver and will generate an example later in a separate pull request.
> What do you think?

Yes, please.

**Comment by mah-eiSmart** [comment]
> > Thus, in case you still not accept the example, I just push the driver and will generate an example later in a separate pull request.
> > What do you think?
> 
> Yes, please.

Hello @MaureenHelm ,
I dropped the example, so only the driver is added.
Review can be done again. Thx for the support.

BR Matthias

File: drivers/sensor/wsen_tids/wsen_tids.c [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+	100,
+	200,
+};
+
+static int tids_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct tids_data *data = dev->data;
+	int16_t raw_temperature;
+
+	if (chan == SENSOR_CHAN_ALL) {
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids.c:35]:
```suggestion
	if ((chan != SENSOR_CHAN_ALL) && (chan != SENSOR_CHAN_AMBIENT_TEMP)) {
```

File: drivers/sensor/wsen_tids/wsen_tids.c [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+		LOG_ERR("Failed to read device ID.");
+		return -EIO;
+	}
+
+	if (device_id != TIDS_DEVICE_ID_VALUE) {
+		LOG_ERR("Invalid device ID 0x%x.", device_id);
+		return -EIO;
+	}
+
+	/* Reset the sensor with an off time of 5 us */
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids.c:152]: Is this time specified in the datasheet or an arbitrarily chosen time?

Comment (mah-eiSmart) [reply drivers/sensor/wsen_tids/wsen_tids.c:152]: No, the time is not specified in the data sheet. We chose that time, and figured out that it's working properly.

**Review by MaureenHelm — APPROVED** [review]

**Comment by mah-eiSmart** [comment]
Hello @gmarull , hello @teburd ,

may I ask for a review please? Thank you!
BR Matthias

**Comment by mah-eiSmart** [comment]
> Hello @gmarull , hello @teburd ,
> 
> may I ask for a review please? Thank you! BR Matthias

Hello @MaureenHelm ,
I don't see any progress here. Sorry for being impatient. Is there any other reviewer?
Thank you for supporting here.
BR Matthias

**Comment by MaureenHelm** [comment]
@teburd please review

**Comment by mah-eiSmart** [comment]
Hello together, hello @MaureenHelm ,

please remove the label "stale" and have a next review to the code to merge my contribution.
Thank you!

File: drivers/sensor/wsen_tids/wsen_tids.c [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "TIDS driver enabled without any devices"
+#endif
+
+/*
+ * Device creation macros
+ */
+
+#define TIDS_DEVICE_INIT(inst)                                        \
+	DEVICE_DT_INST_DEFINE(inst,                                   \
```
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids/wsen_tids.c:195]: Please use `SENSOR_DEVICE_DT_INST_DEFINE`

File: dts/bindings/sensor/we,wsen-tids.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) 2022 Würth Elektronik eiSos GmbH & Co. KG
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Würth Elektronik WSEN-TIDS temperature sensor
+
+compatible: "we,wsen-tids"
+
+include: "i2c-device.yaml"
```
Comment (MaureenHelm) [inline dts/bindings/sensor/we,wsen-tids.yaml:9]:
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Some minor updates needed based on recent changes in main. Please rebase

**Review by MaureenHelm — DISMISSED** [review]

**Comment by mah-eiSmart** [comment]
Hello @teburd,

are you going to review the code?
If not, we need to find another reviewer.
Thank you!
BR Matthias

**Comment by mah-eiSmart** [comment]
Rebased to most recent main because of merge conflicts:
New pull request #54352

**Pull request closed without merging**